src/core/etl/src/Flow/ETL/Row/Entry/ListEntry.php

```diff
@@ -24,26 +21,22 @@ final class ListEntry implements Entry, TypedCollection
 {
     use EntryRef;
 
-    private readonly ListType $type;
-
     /**
      * @param array<T> $value
      *
      * @throws InvalidArgumentException
      */
     public function __construct(
         private readonly string $name,
-        ListElement $element,
-        private readonly array $value
+        private readonly array $value,
+        private readonly ListType $type,
     ) {
         if ('' === $name) {
             throw InvalidArgumentException::because('Entry name cannot be empty');
         }
 
-        $this->type = new ListType($element);
-
-        if (!$this->type->isValid($value)) {
-            throw InvalidArgumentException::because('Expected list of ' . $element->toString() . ' got different types.');
+        if (!$type->isValid($value)) {
+            throw InvalidArgumentException::because('Expected ' . $type->toString() . ' got different types.');
```

Comment:
This does not say too much, I struggled with it before, exception was thrown but I had no idea what is wrong with the provided data structure, what I would do here is at use json encode on the value and add it to the exception message. But probably the best approach would be to use here TypeFactory, generate type and return toString() from it.

Reply:
Improved on list, map & structure, covered with some unit tests.